### ISSUE TITLE
Improve Receive Adapter Performance

### ISSUE THREAD

**Issue by nachocano** [google/knative-gcp#422]
**Problem**
The PullSubscription’s receive adapter pulls messages from a PubSub topic, tries to send them downstream, and then acks or nacks them. Nack-ing will cause immediate retry from PubSub.
The receive adapter currently pulls a single item at a time. This is a core construct, used by every higher level object, thus it’s critical to improve its performance if we want higher throughput.

We need to change the CloudEvents SDK to expose connection knobs to PubSub, and adopt them in the receive adapter. Now that perf tests are reliably running, we can objectively see if this improves things.

**Exit Criteria**
A measurable (binary) test that would indicate that the problem has been resolved. For example, observe that the performance throughput improves

**Time Estimate (optional):** 
How many developer-days do you think this may take to resolve?
< 1w

**Comment by nachocano** [comment]
fyi @yolocs so that you have it on your radar. 
I had some WIP stuff here:
- https://github.com/nachocano/sdk-go/tree/pubsub
- https://github.com/nachocano/knative-gcp/tree/channel_perf

**Comment by nachocano** [comment]
/assign @nlopezgi 

This might already be done... @yolocs added [this change](https://github.com/cloudevents/sdk-go/pull/279) in the CloudEvents SDK to pull stuff from multiple go routines... 
I think is a matter of testing if it works. We might be missing some change in how we construct our pubsub clients in knative-gcp, both in publisher and the receive_adapter (especially in the latter).

@chizhg can we increase the send throughput in our perf tests for PubSub channel?
https://mako.dev/benchmark?benchmark_key=6246352530964480&tseconds=604800

**Comment by ian-mi** [comment]
I'm not sure that this has had the desired effect on performance, the pubsub channel still struggles meet the maximum pace defined by the benchmark. My reading of the pubsub client is that NumGoroutines effectively controls the number of gRPC streams that the client creates and does not limit the concurrency of pubsub message handlers. If this is the case, I believe only a handful of streaming connections (maybe ~num_cpus?) should be necessary to maximize pull throughput so long as we tune the maximum batch size appropriately.

**Comment by nachocano** [comment]
I think those default numbers were mainly borrowed from the tuning of the HTTP client, and as you suggest, may not be correct for the PubSub one. 
Are you interested in taking a closer look at this and see if we can boost the throughput numbers for a single pod?

**Comment by ian-mi** [comment]
Yes, thanks for the additional context, I will look into tuning the receiver parameters. I believe the equivalent receive parameter should be max outstanding messages.

**Comment by yolocs** [comment]
> I'm not sure that this has had the desired effect on performance, the pubsub channel still struggles meet the maximum pace defined by the benchmark. My reading of the pubsub client is that NumGoroutines effectively controls the number of gRPC streams that the client creates and does not limit the concurrency of pubsub message handlers. If this is the case, I believe only a handful of streaming connections (maybe ~num_cpus?) should be necessary to maximize pull throughput so long as we tune the maximum batch size appropriately.

I set the default `NumGoroutines` to 1000 in cloudevents client. Since we block to deliver each message, my understanding is with 1 pod we can only top at 1000 qps. If somehow reality is significantly below that number, we need to find out why.

**Comment by ian-mi** [comment]
While we block delivering messages, the pull receiver will create a goroutine to handle each pubsub message without blocking the pulling thread (see https://github.com/googleapis/google-cloud-go/blob/master/pubsub/subscription.go#L785-L789). As far as I can tell this happens regardless of the value of NumGoroutines which only controls the number for pullers.

**Comment by yolocs** [comment]
Good to know!

**Comment by nachocano** [comment]
synced with @ian-mi offline. He is digging into adding more tracing as that might help out figure out what the bottleneck is. 
@nlopezgi please sync with him before you start with this, so that we can coordinate the effort

/assign @nlopezgi

**Comment by nachocano** [comment]
This will probably be superseded by https://github.com/google/knative-gcp/issues/629. 
Leaving it open for now, but once we have the new binding in and see its improvements, we might be able to close this. Removing the milestone for now

**Comment by grantr** [comment]
@nachocano Any idea what the status of this is now that #629 is in?

**Comment by nachocano** [comment]
I think this is still relevant. The move to ce v2 hasn't been done in sources. I created an issue sometime back to track this: https://github.com/google/knative-gcp/issues/1167
Once that is done, this might be solved.

I suggest to either leave this one open and assign it to Sources area, or just close it and add more details in 1167. The move to ce v2 might actually be a P1.

Thanks for catching this @grantr ! I completely missed it
